### PR TITLE
Adding support for pre and post queries for verifier

### PR DESCRIPTION
**Pull request by drudenko**
@cberner please, have a look. 
Ran verifier locally, for all queries, for which these 4 added db fields are null the verifier just ignores them (as expected). 

I've updated xdb manually for verifier_queries tables.

### CODE REVIEW

**Comment by cberner** [comment]
@drudenko can you try compiling with the Maven command line? It looks like the style checks are failing

**Comment by drudenko** [comment]
My bad, @cberner 

Fixed errors

**Comment by drudenko** [comment]
I'm not sure why continuous-integration is failing and if it's important. Local build is successful :/

File: presto-verifier/src/main/java/com/facebook/presto/verifier/PrestoVerifier.java [hunk 1]
```diff
@@ -62,6 +62,11 @@ public int run(String[] args)
             System.setProperty("config", args[0]);
         }
 
+        if (args.length > 2) {
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/PrestoVerifier.java]: Remove this. Logically separate changes should go in different commits, and I don't think this is how we should pass in the proxy

Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/PrestoVerifier.java]: Thanks, will do

File: presto-verifier/src/main/java/com/facebook/presto/verifier/QueryPairMapper.java [hunk 2]
```diff
@@ -20,12 +21,14 @@
...
 import java.sql.SQLException;
+import java.util.List;
 import java.util.Map;
 
 public class QueryPairMapper
         implements ResultSetMapper<QueryPair>
 {
-    private static final JsonCodec<Map<String, String>> jsonCodec = JsonCodec.mapJsonCodec(String.class, String.class);
+    private static final JsonCodec<Map<String, String>> propertiesJsonCodec = JsonCodec.mapJsonCodec(String.class, String.class);
+    private static final JsonCodec<List<String>> preQueriesJsonCodec = JsonCodec.listJsonCodec(String.class);
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/QueryPairMapper.java]: name this `queriesJsonCodec` since it's used the post queries as well

**Comment by cberner** [comment]
Add release notes (release-0.124.rst if it doesn't already exist) that documents how to migrate the database. See release-0.93.rst for an example

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 3]
```diff
@@ -182,7 +182,60 @@ public boolean isDeterministic()
...
+        for (String query : queryPair.getControl().getPostQueries()) {
+            QueryResult queryResult = executeQueryControl(query);
+            if (queryResult.getState() != State.SUCCESS) {
+                if (testResult.getState() == State.SUCCESS) {
+                    testResult = new QueryResult(State.FAILED_TO_TEARDOWN, null, null, ImmutableList.<List<Object>>of());
+                }
+                return false;
+            }
+        }
+        for (String query : queryPair.getTest().getPreQueries()) {
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: this should be `getPostQueries()`

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 4]
```diff
@@ -182,7 +182,60 @@ public boolean isDeterministic()
...
     {
-        controlResult = executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), controlTimeout, sessionProperties);
+        boolean result = false;
+        if (setupQueries()) {
+            result = validateMainQueries();
+        }
+        return tearDownQuery() && result;
+    }
+
+    private boolean tearDownQuery()
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: I would just name this `tearDown`

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 5]
```diff
@@ -182,7 +182,60 @@ public boolean isDeterministic()
...
+                if (testResult.getState() == State.SUCCESS) {
+                    testResult = new QueryResult(State.FAILED_TO_TEARDOWN, null, null, ImmutableList.<List<Object>>of());
+                }
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean setupQueries()
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: name this `setup`

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 6]
```diff
@@ -182,7 +182,60 @@ public boolean isDeterministic()
 
     private boolean validate()
     {
-        controlResult = executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), controlTimeout, sessionProperties);
+        boolean result = false;
+        if (setupQueries()) {
+            result = validateMainQueries();
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: Put a try-finally around this, and move `tearDown()` call to the finally block

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 7]
```diff
@@ -182,7 +182,60 @@ public boolean isDeterministic()
 
     private boolean validate()
     {
-        controlResult = executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), controlTimeout, sessionProperties);
+        boolean result = false;
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: This needs to be restructured a little. One of the things that the verifier does is re-run the control query a couple times to see if the results are non-deterministic, before declaring that the results are not correct. However, that needs to re-run the setup query as well, otherwise it won't redo the INSERT, it will just re-run the SELECT step. See below for where I would move this to

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 8]
```diff
@@ -253,7 +316,7 @@ public QueryResult getTestResult()
         return testResult;
     }
 
-    private QueryResult executeQuery(String url, String username, String password, Query query, Duration timeout, Map<String, String> sessionProperties)
+    private QueryResult executeQuery(String url, String username, String password, Query query, String sql, Duration timeout, Map<String, String> sessionProperties)
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: I would have both the old method `private QueryResult executeQuery(String url, String username, String password, Query query, Duration timeout, Map<String, String> sessionProperties)` and this one. Then move all the logic to run the pre and post queries into that old method signature, and have it use this one.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 9]
```diff
@@ -238,6 +291,16 @@ else if (controlResult.getState() != State.SUCCESS) {
         return true;
     }
 
+    private QueryResult executeQueryTest(String sql)
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: Probably won't need these methods once you refactor this

**Comment by cberner** [comment]
Let me know when you've addressed the comments and I'll take another look. Also, can you rebase on master? I think that will fix Travis

File: presto-verifier/src/main/java/com/facebook/presto/verifier/QueryPairMapper.java (hunk 2 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/QueryPairMapper.java]: done

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 3 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: Thanks, that was a bug actually

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 4 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 5 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 7 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: Thanks for the comment. I actually understood that logic for a query to be deterministic/non-deterministic. It wasn't just quite clear for me what should be the general behavior for a situation when setup query fails, more specifically, if a control setup query fails? Should we rerun it? Should we execute the teardown before rerunning it? Let's discuss this tomorrow and try to go through all the possible flows and then we'll see if we need the deterministic check for setup()/terdown()

**Comment by drudenko** [comment]
Addressed most of the comment, but we'll have to discuss offline

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 10]
```diff
@@ -182,7 +183,36 @@ public boolean isDeterministic()
 
     private boolean validate()
     {
-        controlResult = executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), controlTimeout, sessionProperties);
+        return validateQueries();
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: inline the `validateQueries()` method since this is the only place it's used

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 11]
```diff
@@ -238,6 +268,36 @@ else if (controlResult.getState() != State.SUCCESS) {
         return true;
     }
 
+    private QueryResult executeQueryTest()
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: this won't run the teardown queries if the setup or test query fails, or if there's an exception. Put the tear down in a finally-block

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 12]
```diff
@@ -238,6 +268,36 @@ else if (controlResult.getState() != State.SUCCESS) {
...
+        }
+        QueryResult queryResult = executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), query.getQuery(), testTimeout, sessionProperties);
+        if (queryResult.getState() != State.SUCCESS) {
+            return queryResult;
+        }
+        QueryResult tearDownResult = tearDown(query, testPostquery -> executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), testPostquery, testTimeout, sessionProperties));
+        return tearDownResult.getState() == State.SUCCESS ? queryResult : tearDownResult;
+    }
+
+    private QueryResult executeQueryControl()
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: this won't run the teardown queries if the setup or test query fails, or if there's an exception. Put the tear down in a finally-block

**Comment by cberner** [comment]
@drudenko couple more comments.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 10 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 12 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: True

**Comment by drudenko** [comment]
Followed the comments and pushed

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 13]
```diff
@@ -205,13 +206,37 @@ else if (controlResult.getState() != State.SUCCESS) {
...
+                return new QueryResult(State.FAILED_TO_SETUP, queryResult.getException(), queryResult.getDuration(), ImmutableList.<List<Object>>of());
+            }
         }
 
-        // check if the control query is deterministic
+        return new QueryResult(State.SUCCESS, null, null, ImmutableList.of());
+    }
+
+    private boolean checkForDeterministicAndRerunTestQueriesIfNeeded()
+    {   // check if the control query is deterministic
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: put comment on a new line

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 14]
```diff
@@ -238,6 +263,52 @@ else if (controlResult.getState() != State.SUCCESS) {
         return true;
     }
 
+    private QueryResult executeQueryTest()
+    {
+        Query query = queryPair.getTest();
+        QueryResult queryResult = new QueryResult(State.INVALID, null, null, ImmutableList.<List<Object>>of());
+        try {
+            //startup
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: space at beginning of comment, to be consistent with others

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 15]
```diff
@@ -238,6 +263,52 @@ else if (controlResult.getState() != State.SUCCESS) {
...
+            queryResult = setup(query, testPrequery -> executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), testPrequery, testTimeout, sessionProperties));
+
+            //if startup is successful -> execute query
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: space at beginning of comment, to be consistent with others

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 16]
```diff
@@ -238,6 +263,52 @@ else if (controlResult.getState() != State.SUCCESS) {
...
+        }
+        return queryResult;
+    }
+
+    private QueryResult executeQueryControl()
+    {
+        Query query = queryPair.getControl();
+        QueryResult queryResult = new QueryResult(State.INVALID, null, null, ImmutableList.<List<Object>>of());
+        try {
+            //startup
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: space at beginning of comment, to be consistent with others

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java [hunk 17]
```diff
@@ -238,6 +263,52 @@ else if (controlResult.getState() != State.SUCCESS) {
...
+            queryResult = setup(query, controlPrequery -> executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), controlPrequery, controlTimeout, sessionProperties));
+
+            //if startup is successful -> execute query
```
Comment (cberner) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: space at beginning of comment, to be consistent with others

**Comment by cberner** [comment]
Looks good, had a couple style comments though

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 13 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 14 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 15 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 16 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

File: presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java (hunk 17 above)
Comment (drudenko) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java]: ok

**Comment by drudenko** [comment]
Fixed style issues
